### ISSUE TITLE
Rule Request: Educated Sorting

### ISSUE THREAD

**Issue by jpsim** [realm/SwiftLint#1038]
There area few arrays in SwiftLint's own source that are alphabetically sorted, like [masterRuleList](https://github.com/realm/SwiftLint/blob/0.15.0/Source/SwiftLintFramework/Models/MasterRuleList.swift#L43-L108) and the [Linux `allTests` extensions](https://github.com/realm/SwiftLint/blob/0.15.0/Tests/SwiftLintFrameworkTests/RulesTests.swift#L347-L409).

The same thing could apply to enum members, dictionaries, constants, etc.

For example, SourceKitten's [SwiftDeclarationKind](https://github.com/jpsim/SourceKitten/blob/0.16.0/Source/SourceKittenFramework/SwiftDeclarationKind.swift) enum members are all sorted.

It'd be awesome if SwiftLint could automatically infer which sets of declarations were _mostly_ sorted and trigger a warning.

For example, if an array of 10 items only has 2 out-of-order elements, odds are that these are just mis-ordered. This likelihood goes up as the number of total items increases and the number of out-of-order elements decreases.

This could be configurable by minimum thresholds, and probably be made correctable in simple cases.

**Comment by JamieEdge** [comment]
Hey, so I've started work on a rule for this (see the related commit above), however when I run unit tests (adding the below code to `RulesTests.swift`), I'm finding the values that are sent in the dictionary in the method call `validateFile(file:kind:dictionary:)` to be very different to 'real' ones.

```
    func testEducatedSorting() {
        verifyRule(EducatedSortingRule.description)
    }
```

Any idea why the body offset and body length come through as out of range values during unit testing? I may just be misunderstanding the purpose of these variables.

```
(lldb) po file.contents.characters.count
59

(lldb) po bodyOffset
93

(lldb) po bodyLength
27
```

**Comment by jpsim** [comment]
The offsets and lengths returned from SourceKit are all in bytes. For multi-byte characters in Strings (e.g. non-ASCII such as emoji), `characters.count != bytes`.

SourceKitten has a number of conversion methods to help with this, most of which are in [`String+SourceKitten.swift`](https://github.com/jpsim/SourceKitten/blob/0.16.0/Source/SourceKittenFramework/String+SourceKitten.swift). Such as [`String.byteRangetoNSRange(start:length:)`](https://github.com/jpsim/SourceKitten/blob/0.16.0/Source/SourceKittenFramework/String+SourceKitten.swift#L268-L276).

**Comment by JamieEdge** [comment]
Thanks @jpsim - that's fixed it now!

**Comment by richellis** [comment]
Would just like to register a vote for this specifically for enum cases and `switch` statements over enum cases. I like the approach in the linked PR for arrays of a minimum number of cases threshold and inferring that the cases are _mostly_ sorted.

**Comment by ZevEisenberg** [comment]
What about a rule where you opt in at the call site, something like:

```swift
let x = [ // swiftlint keep_me_sorted
  "foo",
  "bar", // ⚠️
]

enum Foo { // swiftlint keep_me_sorted
  case foo
  case bar // ⚠️
}
```

naming ideas:

- `keep_me_sorted`
- `auto_sort`
- `sorted_members`
- `alphabetize` (probably not because of `alphabetise` alternate spelling confusion)
